providers/fips/self_test.c

```diff
@@ -311,15 +312,23 @@ int SELF_TEST_post(SELF_TEST_POST_PARAMS *st, int on_demand_test)
     if (ok)
         FIPS_state = FIPS_STATE_RUNNING;
     else
-        ossl_set_error_state();
+        ossl_set_error_state(OSSL_SELF_TEST_TYPE_NONE);
     CRYPTO_THREAD_unlock(self_test_lock);
 
     return ok;
 }
 
-void ossl_set_error_state(void)
+void SELF_TEST_disable_conditional_error_state(void)
 {
-    FIPS_state = FIPS_STATE_ERROR;
+    FIPS_conditional_error_check = 0;
+}
+
+void ossl_set_error_state(const char *type)
+{
+    int cond_test = (type != NULL && strcmp(type, OSSL_SELF_TEST_TYPE_PCT) == 0);
+
+    if (!cond_test || (FIPS_conditional_error_check == 1))
+        FIPS_state = FIPS_STATE_ERROR;
     ERR_raise(ERR_LIB_PROV, PROV_R_FIPS_MODULE_ENTERING_ERROR_STATE);
```

Comment:
This line needs to be inside the if block.  Raising an entering error state when it isn't happening isn't good form.

Reply:
It took too long to figure out that the error state couldn't be turned off by this change :(

Reply:
I dont quite agree.. I would like it to raise an error of some kind..
Add another one if you think it needs it.

Reply:
> It took too long to figure out that the error state couldn't be turned off by this change :(

I dont know what this means.

Reply:
I initially had it just as a int is_conditional that was passed to the function - but that is not very flexible if we decide we need some other check in there at a later point..

Reply:
I've added a different error for the case of a conditional error that doesn't result in the error state.

That sentence was about me needing a break....